### PR TITLE
Documentation and KMeans

### PR DESCRIPTION
**Pull request by ErikNovak**
**Bug Fixes:** 
Issue #418: Categorical feature extraction documentation - Removed the 'values' from the construction documentation.  

Issue #439: Added the two missing optional parameters in the new KMeans constructor, `fitIdx` and `fitStart`. Also fixed the documentation for KMeans constructor parameter and added some new unit tests for KMeans.

### CODE REVIEW

File: src/glib/mine/clustering.h [hunk 1]
```diff
@@ -729,16 +752,66 @@ void TAbsKMeans<TCentroidType>::InitCentroids(TVec<TIntFltKdV>& CentroidVV, cons
...
+template<class TCentroidType>
+void TAbsKMeans<TCentroidType>::InitCentroids(TVec<TIntFltKdV>& CentroidVV, const TVec<TIntFltKdV>& FtrVV) {
+    const int Cols = GetDataCount(FtrVV);
+    // construct the centroid matrix
+    CentroidVV.Gen(Cols);
+    for (int ColN = 0; ColN < Cols; ColN++) {
+        CentroidVV[ColN] = FtrVV[ColN];
+    }
+}
+
```
Comment (rupnikj) [inline src/glib/mine/clustering.h]: These four init methods should reuse code from linalg.h (type transformations) and copy assignment operators.

File: src/glib/mine/clustering.h [hunk 2]
```diff
@@ -922,6 +1000,78 @@ void TDnsKMeans<TCentroidType>::Apply(const TDataType& FtrVV, const int& NInst,
...
+        // recompute the means
+        TAbsKMeans<TCentroidType>::UpdateCentroids(FtrVV, NInst, *AssignIdxVPtr, OnesN, RangeN, TempK, TempDxK, TempKxKSpVV, NormX2, NormC2, AllowEmptyP);
+
+        // swap the old and new assign vectors
+        Temp = AssignIdxVPtr;
+        AssignIdxVPtr = OldAssignIdxVPtr;
+        OldAssignIdxVPtr = Temp;
+    }
+}
```
Comment (rupnikj) [inline src/glib/mine/clustering.h]: Refactorize (copy paste of TDnsKMeans<TCentroidType>::Apply)

File: src/nodejs/analytics/analytics.cpp [hunk 3]
```diff
@@ -2262,38 +2305,15 @@ void TNodeJsKMeans::UpdateParams(const PJsonVal& ParamVal) {
...
-    return ParamVal;
-}
-
 void TNodeJsKMeans::Save(TSOut& SOut) const {
     TInt(Iter).Save(SOut);
     TInt(K).Save(SOut);
     AllowEmptyP.Save(SOut);
     AssignV.Save(SOut);
+	Medoids.Save(SOut);
+	FitIdx.Save(SOut);
```
Comment (rupnikj) [inline src/nodejs/analytics/analytics.cpp]: Use spaces isntead of tabs

File: src/glib/mine/clustering.h [hunk 4]
```diff
@@ -729,16 +752,37 @@ void TAbsKMeans<TCentroidType>::InitCentroids(TVec<TIntFltKdV>& CentroidVV, cons
...
+    CentroidVV = FtrVV;
+}
+
+template<class TCentroidType>
 template<class TDataType>
 inline void TAbsKMeans<TCentroidType>::SelectInitCentroids(const TDataType& FtrVV,
 		const int& K, const int& NInst) {
 
-    EAssertR(NInst >= K, "TStateIdentifier::SelectInitCentroids: The number of initial centroids should be less than the number of data points!");
+	EAssertR(NInst >= K, "TStateIdentifier::SelectInitCentroids: The number of initial centroids should be less than the number of data points!");
```
Comment (rupnikj) [inline src/glib/mine/clustering.h]: spaces instead of tabs

File: src/nodejs/analytics/analytics.cpp [hunk 5]
```diff
@@ -2453,15 +2566,43 @@ void TNodeJsKMeans::TFitTask::Run() {
...
+
+                   TFltVV InitCentroidMat;
+                   int Rows = JsFltVV->Mat.GetRows();
+                   int Cols = JsKMeans->K;
+
+                   InitCentroidMat.Gen(Rows, Cols);
+                   for (int i = 0; i < Cols; i++) {
+                       const int ColN = JsKMeans->FitIdx[i];
+                       for (int RowN = 0; RowN < Rows; RowN++) {
+                           InitCentroidMat.PutXY(RowN, i, JsFltVV->Mat(RowN, ColN));
```
Comment (rupnikj) [inline src/nodejs/analytics/analytics.cpp]: copy assignment

File: src/nodejs/analytics/analytics.cpp [hunk 6]
```diff
@@ -2488,7 +2629,35 @@ void TNodeJsKMeans::TFitTask::Run() {
...
+               }
+               else if (!JsKMeans->FitIdx.Empty()) {
+                   EAssertR(JsKMeans->FitIdx.Len() == JsKMeans->K, "Length of fitIdx must be equal to number of centroids!");
+                   EAssertR(JsKMeans->FitIdx.GetMxVal() < JsSpVV->Mat.Len(), "Maximum index in fitIdx must not be greater to the number of columns!");
+
+                   TFltVV InitCentroidMat;
+                   int Rows = TLinAlgSearch::GetMaxDimIdx(JsSpVV->Mat) + 1;
+                   int Cols = JsKMeans->K;
+
+                   InitCentroidMat.Gen(Rows, Cols);
```
Comment (rupnikj) [inline src/nodejs/analytics/analytics.cpp]: copy assignment

File: src/nodejs/analytics/analytics.cpp [hunk 7]
```diff
@@ -2518,11 +2687,39 @@ void TNodeJsKMeans::TFitTask::Run() {
...
+                   EAssertR(JsKMeans->FitIdx.GetMxVal() < JsFltVV->Mat.GetCols(), "Maximum index in fitIdx must not be greater to the number of columns!");
+
+                   TVec<TIntFltKdV> InitCentroidMat;
+                   int Rows = JsFltVV->Mat.GetRows();
+                   int K = JsKMeans->K;
+                   InitCentroidMat.Gen(K);
+                   for (int i = 0; i < K; i++) {
+                       const int ColN = JsKMeans->FitIdx[i];
+                       for (int RowN = 0; RowN < Rows; RowN++) {
+                           if (JsFltVV->Mat(RowN, ColN) != 0.0) { InitCentroidMat[i].Add(TIntFltKd(RowN, JsFltVV->Mat(RowN, ColN))); }
```
Comment (rupnikj) [inline src/nodejs/analytics/analytics.cpp]: use linalg type transofrmation functions

File: src/nodejs/analytics/analytics.cpp [hunk 8]
```diff
@@ -2549,7 +2746,32 @@ void TNodeJsKMeans::TFitTask::Run() {
...
+               }
+               else if (!JsKMeans->FitIdx.Empty()) {
+                   EAssertR(JsKMeans->FitIdx.Len() == JsKMeans->K, "Length of fitIdx must be equal to number of centroids!");
+                   EAssertR(JsKMeans->FitIdx.GetMxVal() < JsSpVV->Mat.Len(), "Maximum index in fitIdx must not be greater to the number of columns!");
+
+                   TVec<TIntFltKdV> InitCentroidMat;
+                   int Dim = JsKMeans->K;
+
+                   InitCentroidMat.Gen(Dim);
+                   for (int i = 0; i < Dim; i++) {
```
Comment (rupnikj) [inline src/nodejs/analytics/analytics.cpp]: fix

File: src/glib/mine/clustering.h (hunk 1 above)
Comment (ErikNovak) [inline src/glib/mine/clustering.h]: The init methods now use the code from linalg.h

File: src/glib/mine/clustering.h (hunk 2 above)
Comment (ErikNovak) [inline src/glib/mine/clustering.h]: Refactorized the code.

**Pull request merged**